### PR TITLE
Reimplement jinja[spacing] to avoid use of regex

### PR DESCRIPTION
**Pull request by ssbarnea**
This change rewrites jinja2 rule to make it use jinaja lexer and
avoid use of regexes for parsing, as they cause too many false
positives.

Fixes: #2301
Fixes: #2285
Fixes: #2241
Fixes: #2209
Fixes: #2208
Fixes: #2120

### CODE REVIEW

**Review by cognifloyd — APPROVED** [review]
fascinating use of the jinja lexer.
In the lexer-related work I've been doing, I've been using `TOKEN_*` constants from jinja:
https://github.com/pallets/jinja/blob/main/src/jinja2/lexer.py#L61-L109

Typically I do `from jinja2 import lexer as j2tokens` and then use `j2tokens.TOKEN_OPERATOR` or similar. What do you think of that idiom?

File: src/ansiblelint/rules/jinja.py [hunk 1]
```diff
@@ -94,6 +86,83 @@ def matchyaml(self, file: Lintable) -> List[MatchError]:
...
+                    # breakpoint()
+                    value = " "
+
+                tokens.append((lineno, token_type, value))
+                prev_token_type = token_type
+                prev_value = value
+
+            reformatted = "".join([value for _, _, value in tokens])
+            details = (
+                f"Jinja2 template could better be rewritten as `{reformatted}`."
```
[the remark this answers is not in the document]
Comment (ssbarnea) [reply src/ansiblelint/rules/jinja.py:156]: Tx. FYI, I found other issues and I am rewriting it, while keeping the tests and likely adding few more.

**Comment by ssbarnea** [comment]
@drybjed I am not sure how well you will take this but this change seems to uncover at least one genuine jinja2 syntax error in debops repository, but you might find yourself bit overwhelmed by the amount of feedback linter is providing:



Take it easy! And to quote from [thread](https://github.com/ansible/ansible-lint/discussions/2205),... I hope you get paid by the hour... ;)

**Comment by ssbarnea** [comment]
@cognifloyd There are 4 reformatting test that are failing which I do not know how to fix. If you can have a look it would be great. That PR took all my weekend and I might really benefit from another pair of eyes to deal with these last ones.

All these tests were results of different errors I found while testing formatting of jinja2 expression on all eco repositories, so we should be pleased about the results.

Few minutes ago I got another idea,... what if I would try to use black itself to format the snippets, assuming that it might be used programately. Still, doing that would not be a joy as I would have to identify blocks of tokens and recombine them before I feed them to black,.. so maybe not a very good idea.

File: src/ansiblelint/rules/jinja.py [hunk 2]
```diff
@@ -131,58 +344,175 @@ def test_var_spacing(
...
+            pytest.param(
+                # "{% if a|int <= 8 -%} iptables {%- else -%} iptables-nft {%- endif %}",
+                "{% if a|int <= 8 -%} iptables {%- else -%} iptables-nft {%- endif %}",
+                "{% if a | int <= 8 -%} iptables {%- else -%} iptables-nft {%- endif %}",
+                "spacing",
+                id="29",
+            ),
+            pytest.param(
+                # "- 2" -> "-2", minus does not get separated when there is no left side
+                "{{ - 2 }}",
```
Comment (cognifloyd) [inline src/ansiblelint/rules/jinja.py:485]: The AST calls this an Unary node. I looked through the parser, and I don't see an easy way to identify this based only on the lexer tokens.

So maybe track the previous non-whitespace token. See if the token before the `-` or `+` operator is also an operator (one of https://github.com/pallets/jinja/blob/main/src/jinja2/lexer.py#L112-L139 except for right tokens: `]`, `)`, `}`) or if the previous token is a variable start token `{{`.

File: src/ansiblelint/rules/jinja.py [hunk 3]
```diff
@@ -131,58 +344,175 @@ def test_var_spacing(
...
+            pytest.param(
+                # array ranges do not have space added
+                "{{ foo[2:4] }}",
+                "{{ foo[2:4] }}",
+                "spacing",
+                id="32",
+            ),
+            pytest.param(
+                # array ranges have the extra space removed
+                "{{ foo[2: 4] }}",
```
Comment (cognifloyd) [inline src/ansiblelint/rules/jinja.py:506]: In the AST, this is a Slice node.

You will need to track the state to successfully identify this. If within `[` skip the spaces. Maybe have a stack var, and every time you enter a left operator, put it on the stack (a list), then pop it once you hit the closing token. So, if within `{}`, `:` gets a space after. If within `[]` no spaces around `:`.

**Review by cognifloyd — COMMENTED** [review]
Ew. Yes those are difficult. Here are some ideas to hopefully give you an idea of what to look at next.

**Comment by drybjed** [comment]
@ssbarnea Thanks for the heads up. I'll test the proposed changes and make fixes ASAP.

File: src/ansiblelint/rules/jinja.py [hunk 4]
```diff
@@ -131,58 +357,175 @@ def test_var_spacing(
...
+                "{{ foo[2: 4] }}",
```
Comment (cognifloyd) [inline src/ansiblelint/rules/jinja.py:520]:
```suggestion
                "{{ foo[2:4] }}",
```

File: src/ansiblelint/rules/jinja.py (hunk 3 above)
Comment (ssbarnea) [reply src/ansiblelint/rules/jinja.py:523]: No AST/black for the moment, likely to open another can of worms. Time will tell if we need them or not.

**Comment by drybjed** [comment]
> @drybjed I am not sure how well you will take this but this change seems to uncover at least one genuine jinja2 syntax error in debops repository, but you might find yourself bit overwhelmed by the amount of feedback linter is providing:
> 
> 
> 
> Take it easy! And to quote from [thread](https://github.com/ansible/ansible-lint/discussions/2205),... I hope you get paid by the hour... ;)

@ssbarnea I'm currently working on fixing all these warnings on `ansible-lint` v6.5.0 so that DebOps can be used to fix false-positives in v6.5.1. I wish for one thing - separate Jinja identation into its own rule so that it could be ignored if desired. There are small snippets of Jinja where I had to drop idented `if`/`endif` statements, and that's alright with me, but for larger pieces of Jinja, for example [more extensive templates](https://github.com/debops/debops/blob/master/ansible/roles/journald/templates/etc/systemd/journald.conf.d/ansible.conf.j2), please please please let me keep identation - it makes things much more readable.

**Comment by drybjed** [comment]
@ssbarnea All issues in DebOps have been [fixed](https://github.com/debops/debops/pull/2185) in the `master` branch. There are some false-positives though on the latest `ansible-lint` release.

**Comment by ssbarnea** [comment]
@drybjed Thanks. I got some progress today on them but again failed to pass all tests. That jinja spacing is a bit of a can of worms.

I already introduce a condition to skip the spacing checks when newlines are found inside the jinja blocks, basically because we assume user added them for indentation purposes. The scope of checking spaces was more towards simple blocks.

**Pull request merged**